File: pkg/cmd/execution_argv_test.go
// @custom start
// Regression coverage for handwritten execution argument forwarding.

package cmd

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"os/exec"
	"path/filepath"
	"reflect"
	"runtime"
	"strconv"
	"strings"
	"sync/atomic"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestInvariantExecutionArgumentsReachAPIUnchanged checks the built CLI against
// a local HTTP server. It covers literal arguments, existing input formats and
// subcommands, help output, and rejection before any request is sent.
func TestInvariantExecutionArgumentsReachAPIUnchanged(t *testing.T) {
	// A real binary exercises command routing and request encoding together.
	binary := filepath.Join(t.TempDir(), "dedalus")
	if runtime.GOOS == "windows" {
		binary += ".exe"
	}
	output, err := exec.Command("go", "build", "-o", binary, "../../cmd/dedalus").CombinedOutput()
	require.NoError(t, err, "%s", output)

	// An existing file makes accidental @file expansion observable in the body.
	localFile := filepath.Join(t.TempDir(), "literal.txt")
	require.NoError(t, os.WriteFile(localFile, []byte("must not be uploaded"), 0600))
	// These values must survive without shell interpretation or flag parsing.
	argv := []string{"echo", "hello world", "", "--help", "--", "a\"b", "$HOME", "&&", "@" + localFile, `\@literal`}

	// Both resource names support explicit create and the shorthand form.
	for _, resource := range []string{"executions", "exec"} {
		for _, create := range []bool{false, true} {
			t.Run(resource+"/create="+strconv.FormatBool(create), func(t *testing.T) {
				// Buffer the body so the handler can reply before the CLI exits.
				requests := make(chan map[string]any, 1)
				server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
					assert.Equal(t, "POST", r.Method)
					assert.Equal(t, "/v1/machines/dm-test/executions", r.URL.Path)
					var body map[string]any
					assert.NoError(t, json.NewDecoder(r.Body).Decode(&body))
					requests <- body
					w.Header().Set("Content-Type", "application/json")
					_, _ = w.Write([]byte(`{"execution_id":"ex-test","status":"queued"}`))
				}))
				defer server.Close()
				args := []string{"--base-url", server.URL, "--api-key", "test", "machines", resource}
				if create {
					args = append(args, "create")
				}
				args = append(args, "--machine-id", "dm-test", "--cwd", "/tmp", "--env", `{"MODE":"test"}`, "--stdin", "input", "--timeout-ms", "1000", "--")
				args = append(args, argv...)
				output, err := exec.Command(binary, args...).CombinedOutput()
				require.NoError(t, err, "%s", output)
				select {
				case body := <-requests:
					// Decode the expected slice too: JSON arrays in map[string]any
					// have type []any, so []string would fail a type-sensitive comparison.
					encoded, err := json.Marshal(argv)
					require.NoError(t, err)
					var want any
					require.NoError(t, json.Unmarshal(encoded, &want))
					require.Equal(t, want, body["command"])
					require.Equal(t, "/tmp", body["cwd"])
					require.Equal(t, map[string]any{"MODE": "test"}, body["env"])
					require.Equal(t, "input", body["stdin"])
					require.Equal(t, float64(1000), body["timeout_ms"])
				default:
					t.Fatal("no execution request received")
				}
			})
		}
	}

	// The alias preserves JSON and piped input plus every generated subcommand.
	// A command named list after -- must still create a remote execution.
	for _, test := range []struct {
		name    string
		args    []string
		stdin   string
		command []string
		method  string
		suffix  string
	}{
		{name: "subcommand name is literal", args: []string{"exec", "--machine-id", "dm-test", "--", "list"}, command: []string{"list"}},
		{name: "explicit remote shell", args: []string{"exec", "--machine-id", "dm-test", "--", "sh", "-c", `echo "hello world" && ls -la`}, command: []string{"sh", "-c", `echo "hello world" && ls -la`}},
		{name: "json command", args: []string{"exec", "create", "--machine-id", "dm-test", "--command", `["echo","hello world"]`}, command: []string{"echo", "hello world"}},
		{name: "piped command", args: []string{"exec", "create"}, stdin: `{"machine_id":"dm-test","command":["echo","hello world"]}`, command: []string{"echo", "hello world"}},
		{name: "retrieve alias", args: []string{"exec", "retrieve", "--machine-id", "dm-test", "--execution-id", "ex-test"}, method: "GET", suffix: "/ex-test"},
		{name: "output alias", args: []string{"exec", "output", "--machine-id", "dm-test", "--execution-id", "ex-test"}, method: "GET", suffix: "/ex-test/output"},
		{name: "events alias", args: []string{"exec", "events", "--machine-id", "dm-test", "--execution-id", "ex-test"}, method: "GET", suffix: "/ex-test/events"},
		{name: "list alias", args: []string{"exec", "list", "--machine-id", "dm-test"}, method: "GET"},
		{name: "delete alias", args: []string{"exec", "delete", "--machine-id", "dm-test", "--execution-id", "ex-test"}, method: "DELETE", suffix: "/ex-test"},
	} {
		t.Run(test.name, func(t *testing.T) {
			requests := make(chan *http.Request, 1)
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				if test.command != nil {
					var body struct {
						Command []string `json:"command"`
					}
					if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
						t.Error(err)
					}
					if !reflect.DeepEqual(test.command, body.Command) {
						t.Errorf("command = %q, want %q", body.Command, test.command)
					}
				}
				requests <- r
				w.Header().Set("Content-Type", "application/json")
				_, _ = w.Write([]byte(`{"execution_id":"ex-test","status":"queued","items":[]}`))
			}))
			defer server.Close()
			args := append([]string{"--base-url", server.URL, "--api-key", "test", "machines"}, test.args...)
			run := exec.Command(binary, args...)
			run.Stdin = strings.NewReader(test.stdin)
			output, err := run.CombinedOutput()
			require.NoError(t, err, "%s", output)
			select {
			case request := <-requests:
				method := test.method
				if method == "" {
					method = "POST"
				}
				require.Equal(t, method, request.Method)
				require.Equal(t, "/v1/machines/dm-test/executions"+test.suffix, request.URL.Path)
			default:
				t.Fatal("no execution request received")
			}
		})
	}

	// Help must remain reachable at both the resource and create levels.
	for _, args := range [][]string{{"exec", "--help"}, {"exec", "create", "--help"}} {
		output, err := exec.Command(binary, append([]string{"machines"}, args...)...).CombinedOutput()
		require.NoError(t, err, "%s", output)
		require.Contains(t, string(output), "machine-id")
	}

	// Invalid input must fail locally. A nonzero exit alone would also allow
	// a server-side rejection, so the atomic counter checks for zero requests.
	for _, args := range [][]string{
		{"exec", "--machine-id", "dm-test", "--"},
		{"exec", "--machine-id", "dm-test", "--", ""},
		{"exec", "--machine-id", "dm-test", "--command", `["echo"]`, "--", "pwd"},
		{"exec", "--", "echo"},
		{"exec", "typo"},
	} {
		t.Run("reject/"+strings.Join(args, " "), func(t *testing.T) {
			var requests atomic.Int32
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				requests.Add(1)
				w.Header().Set("Content-Type", "application/json")
				_, _ = w.Write([]byte(`{"execution_id":"unexpected"}`))
			}))
			defer server.Close()
			output, err := exec.Command(binary, append([]string{"--base-url", server.URL, "--api-key", "test", "machines"}, args...)...).CombinedOutput()
			require.Error(t, err, "%s", output)
			require.Zero(t, requests.Load(), "invalid input must not send an API request")
		})
	}
}

// @custom end

File: tests/execution-argv.test.mjs
import assert from 'node:assert/strict'
import { spawn } from 'node:child_process'
import { resolve as resolvePath } from 'node:path'
import { createServer } from 'node:http'
import { test } from 'node:test'

const run = (args, baseURL) => new Promise((resolve, reject) => {
  const child = spawn(process.execPath, [resolvePath(process.env.FEEDBACK_PACKAGE_ROOT ?? '.', 'dist/esm/bin.js'), '--base-url', baseURL,
    '--api-key', 'test', '--max-retries', '0', ...args], { stdio: ['pipe', 'pipe', 'pipe'], env: { PATH: process.env.PATH } })
  let stdout = '', stderr = ''
  child.stdout.on('data', (data) => { stdout += data })
  child.stderr.on('data', (data) => { stderr += data })
  child.on('error', reject)
  child.on('close', (status) => resolve({ status, stdout, stderr }))
  child.stdin.end()
})

const withAPI = async (action) => {
  const requests = []
  const server = createServer(async (request, response) => {
    let source = ''
    for await (const chunk of request) source += chunk
    requests.push({ method: request.method, url: request.url, body: source ? JSON.parse(source) : undefined })
    response.setHeader('content-type', 'application/json')
    response.end(JSON.stringify(request.method === 'GET' ? { items: [] } : { execution_id: 'execution' }))
  })
  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve))
  try { await action(`http://127.0.0.1:${server.address().port}`, requests) }
  finally { await new Promise((resolve) => server.close(resolve)) }
}

const resources = [['machines', 'exec'], ['machines', 'executions'], ['machines:exec'], ['machines:executions']]

test('invariant_execution_argv_reaches_the_wire_literally_for_every_resource_spelling', async () => {
  await withAPI(async (baseURL, requests) => {
    const argv = ['printf', '@does-not-exist', '\\@literal', '--help', '--command', 'two words', '', 'null', '123', '--']
    for (const resource of resources) {
      for (const verb of [[], ['create']]) {
        const result = await run([...resource, ...verb, '--machine-id', 'machine', '--cwd', '/tmp', '--', ...argv], baseURL)
        assert.equal(result.status, 0, result.stderr)
        assert.deepEqual(requests.at(-1), { method: 'POST', url: '/v1/machines/machine/executions', body: { command: argv, cwd: '/tmp' } })
      }
    }
  })
})

test('invariant_invalid_execution_syntax_never_sends_a_request', async () => {
  await withAPI(async (baseURL, requests) => {
    for (const suffix of [
      ['--command', '["echo"]', '--', 'other'],
      ['--', '', 'argument'],
      ['echo'],
      ['echo', '--', 'argument'],
    ]) {
      const result = await run(['machines', 'exec', 'create', '--machine-id', 'machine', ...suffix], baseURL)
      assert.equal(result.status, 2, result.stderr)
    }
    assert.deepEqual(requests, [])
  })
})

test('invariant_existing_execution_flags_and_management_commands_remain_available', async () => {
  await withAPI(async (baseURL, requests) => {
    for (const resource of resources) {
      const create = await run([...resource, 'create', '--machine-id', 'machine', '--command', '["echo","hello"]'], baseURL)
      assert.equal(create.status, 0, create.stderr)
      assert.deepEqual(requests.at(-1).body, { command: ['echo', 'hello'] })
      const list = await run([...resource, 'list', '--machine-id', 'machine'], baseURL)
      assert.equal(list.status, 0, list.stderr)
      assert.equal(requests.at(-1).method, 'GET')
      const help = await run([...resource, '--help'], baseURL)
      assert.equal(help.status, 0, help.stderr)
      assert.match(help.stdout, /create.*argv/u)
    }
  })
})
